File: Java/src/NapakalakiGame/CultistPlayer.java
/*
 * Autores:
 * Guillermo Montes Martos
 * Manuel Ruiz López
 *
 */
package NapakalakiGame;

import java.util.Random;

public class CultistPlayer extends Player{
    
    private static int totalCultistPlayers = 0;
    private Cultist myCultistCard;
    
    
    // Constructor
    
    public CultistPlayer(Player p, Cultist c){
        super(p);
        myCultistCard = c;
        totalCultistPlayers++;
    }
        
    @Override
    public int getCombatLevel(){
        return ( (int) (super.getCombatLevel() * 1.2 + myCultistCard.getGainedLevels() * totalCultistPlayers) );
    }
    
    @Override
    protected int getOponentLevel(Monster m){
        return m.getCombatLevelAgainstCultistPlayer();
    }
    
    @Override
    protected boolean shouldConvert (){
        return false;
    }

    @Override
    protected Treasure giveMeATreasure (){
        Random n = new Random();
        return getVisibleTreasures().remove(n.nextInt(getVisibleTreasures().size()));
    }
    
    @Override
    protected boolean canYouGiveMeATreasure(){
        return !getVisibleTreasures().isEmpty();
    }
    
    public static int getTotalCultistPlayers(){
        return totalCultistPlayers;
    }
    
}

File: Ruby/lib/CultistPlayer.rb
#encoding: utf-8

# Autores:
# 
#     Guillermo Montes Martos
#     Manuel Ruiz López
#

require_relative 'Cultist.rb'
require_relative 'Player.rb'


module NapakalakiGame
  class CultistPlayer < Player
    
    @@totalCultistPlayers = 0
    
    def initialize(p, c)
      super(p.name)
      @level = p.level
      @dead = p.dead
      @canISteal = p.canISteal
      @hiddenTreasures = p.hiddenTreasures
      @visibleTreasures = p.visibleTreasures
      @enemy = p.enemy
      @pendingBadConsequence = p.pendingBadConsequence
      @myCultistCard = c
      @totalCultistPlayers += 1
    end
    
    protected
    def getOponentLevel(m)
      m.getCombatLevelAgainstCultistPlayer
    end
    
    def shouldConvert()
      false
    end
    
    def getCombatLevel()
      (super * 1.2 + @myCultistCard.gainedLevels).floor
    end
    
    def giveMeATreasure()
      @visibleTreasures.delete_at(rand(@visibleTreasures.length))
    end
    
    def canYouGiveMeATreasure()
      !@visibleTreasures.empty?
    end
    
  end
end
